synapse/handlers/message.py

```diff
@@ -1244,7 +1235,7 @@ async def create_new_client_event(
             context.app_service = requester.app_service
 
         res, new_content = await self.third_party_event_rules.check_event_allowed(
-            event, context
+            event, context, for_batch=for_batch, state_map=state_map
         )
```

Comment:
I wonder if we should move this to `handle_new_client_event` (where we do things like check event auth), so that we can avoid having to create an `EventContext` in this function entirely?

Reply:
To make sure I'm following, in that case, `create_new_client_event` would no longer create an event context for either batched or non-batched events, and whoever is creating the event/s would be responsible for calling `compute_event_context` + batched variant to then create the event contexts after the event/s are created?

Reply:
Yup, `create_new_client_event` would only return an `Event`. We could move creating the event context into `handle_new_client_event`

Reply:
Got it, makes sense.

Reply:
Before I go down the road to of a major refactor, there are a number of places specifically in the federation code where create_new_client_event is called and the context is immediately used, for example in the code to exchange a third party [invite](https://github.com/matrix-org/synapse/blob/b2c2b030798d0e74d3bf1afb4726465b53620638/synapse/handlers/federation.py#L1317) we use the context to add a display name to the third party invite and check the signature all before the event and context are sent on to handle_new_client_event. In these cases would I create the context in that code before handle_new_client_event and modify handle_new_client_event to deal with the case where some events may already have had their contexts created?

Reply:
Also the work to move creating event contexts in `handle_new_client_events` is a little extensive, does it make sense to break this out into another pr?